### PR TITLE
Add walsh(n) function that returns a Walsh matrix

### PR DESCRIPTION
**Pull request by syoshida1983**
I added a function `walsh(n)` that returns a Walsh matrix by bit-reversal permutation of a Hadamard matrix. The following package provides a similar function but merges it into Hadamard.jl would be helpful.
https://github.com/syoshida1983/Walsh.jl

### CODE REVIEW

File: src/Hadamard.jl [hunk 1]
```diff
@@ -368,6 +368,21 @@ function hadamard(n::Integer)
     return H
 end
 
+"""
+    walsh(n)
+
+Return a Walsh matrix of order `n`. by bit-reversal permutation of a Hadamard matrix.
+"""
+function walsh(n::Integer)
```
Comment (stevengj) [inline src/Hadamard.jl:376]:
```suggestion
walsh(n::Integer) = walsh(Int(n))

function walsh(n::Int)
    ispow2(n) || throw(ArgumentError("n=$n is not a power of two"))
```

File: src/Hadamard.jl [hunk 2]
```diff
@@ -368,6 +368,21 @@ function hadamard(n::Integer)
...
+    m = Int(round(log2(n))) + 1 # number of bits to represent the index
```
Comment (stevengj) [inline src/Hadamard.jl:377]:
```suggestion
    m = trailing_zeros(n) + 1 # number of bits to represent the index
```

File: src/Hadamard.jl [hunk 3]
```diff
@@ -368,6 +368,21 @@ function hadamard(n::Integer)
...
+    e = sizeof(Int)*8 - m       # number of extra bits
+    i = 0:n-1                   # Hadamard index
+    j = bitreverse.(i) .>> e    # bit reversing of the binary index
+    j = j .⊻ (j .>> 1)          # binary sequency index
+
+    return hadamard(n)[j.+1,:]
```
Comment (stevengj) [inline src/Hadamard.jl:383]: This eliminates a couple of temporary arrays:
```suggestion
    j = [ let b = bitreverse(i) >> e # bit-reverse the binary index (trailing m bits)
              j = b ⊻ (b >> 1) # binary sequency index
              j + 1 # 1-based index
          end
          for i in 0:n-1 ] 

    return hadamard(n)[j, :]
```

**Comment by stevengj** [comment]
Also needs some tests, e.g.
```jl
@test walsh(8) == [1 1 1 1 1 1 1 1; 1 1 1 1 -1 -1 -1 -1; 1 1 -1 -1 -1 -1 1 1; 1 1 -1 -1 1 1 -1 -1; 1 -1 -1 1 1 -1 -1 1; 1 -1 -1 1 -1 1 1 -1; 1 -1 1 -1 -1 1 -1 1; 1 -1 1 -1 1 -1 1 -1]
@test_throws ArgumentError walsh(9)
```

**Comment by syoshida1983** [comment]
I updated README.md.

File: README.md [hunk 4]
```diff
@@ -68,6 +68,11 @@ choice is equivalent to `ifwht_natural(eye(n), 1)`.
 You can pretty-print a Hadamard matrix as a table of `+` and `-`
 (characters indicating the signs of the entries) via `Hadamard.printsigns`, e.g. `Hadamard.printsigns(hadamard(28))` for the 28×28 Hadamard matrix.
 
+You can also obtain a Walsh matrix (sequency-ordered Hadamard matrix) of
+order `n` by using the function `walsh(n)`; the order `n` *must* be powers
```
Comment (stevengj) [inline README.md:72]:
```suggestion
order `n` by using the function `walsh(n)`; the order `n` *must* be a power
```

**Pull request merged**